### PR TITLE
Add Z-Loss support, use flash_attn CrossEntropy, don't recalculate CrossEntropy metric if unneeded

### PR DESCRIPTION
**Pull request by warner-benjamin**
This PR adds support for Z-Loss by using Flash Attention's CrossEntropy loss function and uses FA's CE by default. It also doesn't recalculate the cross-entropy loss for metrics by default but logs the loss as a metric.

The combined changes result in a ~8.5% training speed increase and ~23% memory reduction (I assume because FA CE doesn't materialize the entire matrix).

If you want to calculate CrossEntropy loss as a metric (perhaps when using label smoothing, but want to record non-label smoothing loss), it appears to at max 1 percent slower.

### CODE REVIEW

**Comment by warner-benjamin** [comment]
Merging since I've been training with this for the past couple days without issue.
